### PR TITLE
Fix logprob bug

### PR DESCRIPTION
**Pull request by AlexandraVolokhova**
There is a bug introduced by the logprobs efficiency PR: runs are crushing with gflownet.collect_reversed_logprobs=False and with both forward and backward trajectories present in the training batch. This PR fixes this bug.

### CODE REVIEW

File: gflownet/utils/batch.py [hunk 1]
```diff
@@ -484,6 +486,7 @@ def get_states(
         policy: Optional[bool] = False,
         proxy: Optional[bool] = False,
         force_recompute: Optional[bool] = False,
+        index: Optional[Union[List, Tuple, TensorType, npt.NDArray]] = None,
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:489]: I would call this variable `indices` (plural) instead of `index` (singular).

File: gflownet/utils/common.py [hunk 2]
```diff
@@ -773,3 +774,40 @@ def example_documented_function(arg1, arg2):
     if arg1 == arg2:
         raise ValueError("arg1 must not be equal to arg2")
     return True
+
+
+def index_select(
```
Comment (alexhernandezgarcia) [inline gflownet/utils/common.py:779]: I would call this method `select_indices()`.

File: gflownet/utils/common.py [hunk 3]
```diff
@@ -773,3 +774,40 @@ def example_documented_function(arg1, arg2):
...
+        result = [iterable[idx] for idx in index]
+        if isinstance(iterable, tuple):
+            result = tuple(result)
+        return result
+    elif torch.is_tensor(iterable) or isinstance(iterable, np.ndarray):
+        if isinstance(index, tuple):
+            index = list(index)
+        return iterable[index]
+    else:
+        raise Exception(f"Cannon select elements from {type(iterable)}")
```
Comment (alexhernandezgarcia) [inline gflownet/utils/common.py:813]: Typo:
```suggestion
        raise Exception(f"Cannot select elements from {type(iterable)}")
```

File: gflownet/utils/batch.py [hunk 4]
```diff
@@ -668,22 +686,51 @@ def get_logprobs(
...
+            for idx, lp in enumerate(logprobs):
+                if lp is None:
+                    logprobs_valid[idx] = False
+                    logprobs_clean.append(
+                        tfloat(0.0, device=self.device, float_type=self.float)
+                    )
+                else:
+                    logprobs_clean.append(lp)
+        else:
+            logprobs_clean = logprobs
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:728]: I think there are a bunch of inefficient things here:
- Two for loops that can be avoided
- Tensor creation scalar by scalar
I believe all this can be improved through indexing with the help of numpy, which allows creating an array even though it's got None's in it. I already used this trick [here](https://github.com/alexhernandezgarcia/gflownet/blob/main/gflownet/gflownet.py#L407).
I'm making a suggestion directly here but I'm not testing it so even the syntax may be wrong. It's just for illustration, but please check it yourself.  
```suggestion
        logprobs_clean = np.array(logprobs)
        logprobs_clean[logprobs_clean == None] = 0.0
```

File: gflownet/utils/batch.py [hunk 5]
```diff
@@ -693,7 +740,10 @@ def get_done(self) -> TensorType["n_states"]:
 
     # TODO: check availability one by one as in get_masks
     def get_parents(
-        self, policy: Optional[bool] = False, force_recompute: Optional[bool] = False
+        self,
+        policy: Optional[bool] = False,
+        force_recompute: Optional[bool] = False,
+        index: Optional[Union[List, Tuple, TensorType, npt.NDArray]] = None,
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:746]: Same, `indices` instead of `index` (suggestion).

File: gflownet/utils/batch.py (hunk 4 above)
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py:728]: The problem is that in the general case, some of the elements in the `logprobs ` list are tensors with attached gradients to them and some others are Nones. So, I cannot convert the whole `logprobs ` into a numpy array as I will lose gradients. I also cannot convert it into a tensor because torch throws an error because of the Nones (you cannot do `torch.tensor([tensor, None, tensor, None]`). Therefore, iterating over the list and converting one-by-one is the only possible solution here

File: gflownet/utils/common.py (hunk 2 above)
Comment (AlexandraVolokhova) [reply gflownet/utils/common.py]: Changed here: https://github.com/alexhernandezgarcia/gflownet/pull/396/commits/31b05cc805fa7377036c110e4c3a34e70ea133bb

File: gflownet/utils/batch.py (hunk 5 above)
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py]: Changed here: https://github.com/alexhernandezgarcia/gflownet/pull/396/commits/c6adc2638656de26f65b84534a3b283c3ab51f7d

File: gflownet/utils/batch.py [hunk 6]
```diff
@@ -1019,6 +1080,7 @@ def _compute_masks_forward(self):
     def get_masks_backward(
         self,
         force_recompute: bool = False,
+        index: Optional[Union[List, Tuple, TensorType, npt.NDArray]] = None,
```
[the remark this answers is not in the document]
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py]: Changed here: https://github.com/alexhernandezgarcia/gflownet/pull/396/commits/c6adc2638656de26f65b84534a3b283c3ab51f7d

File: gflownet/utils/batch.py [hunk 7]
```diff
@@ -668,22 +686,51 @@ def get_logprobs(
...
-        return logprobs, valids
+        )
+        self.logprobs_backward, self.logprobs_backward_valid = (
+            self._logprobs_to_clean_tensor(
+                self.logprobs_backward, self.logprobs_backward_valid
+            )
+        )
+        self._logprobs_available = True
+
+    def _logprobs_to_clean_tensor(self, logprobs, logprobs_valid):
```
[the remark this answers is not in the document]
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py]: Done here: https://github.com/alexhernandezgarcia/gflownet/pull/396/commits/e9ee12d1c898436488a2be646a6fdc8722ec2d24

File: gflownet/utils/batch.py (hunk 4 above)
Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py:748]: Oh I see! I had not thought of the gradients, but now I think about it, I'm surprised this works at all because the line below would destroy the gradients too!
```
logprobs_clean = tfloat(
    logprobs_clean, float_type=self.float, device=self.device
)
```

Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py:748]: With this in mind, then how about the following:
```
logprobs_clean = [logp if logp not None else 0.0 for logp in logprobs]
```

File: gflownet/utils/common.py (hunk 2 above)
Comment (alexhernandezgarcia) [reply gflownet/utils/common.py]: Thanks for making this change!

File: gflownet/utils/batch.py [hunk 8]
```diff
@@ -668,22 +686,71 @@ def get_logprobs(
...
+            logprob (True) or not (False). When the flag is False, the corresponding logprob
+            is either None or torch.tensor(0.0) detached from the computational graph.
+
+        Returns
+        -------
+        (logprobs_clean, logprobs_valid): a tuple of 1-d tensors
+            - logprobs_clean is a 1-d float tensor with logprobs, where all valid logprobs
+            remain connected to their computational graphs and all non-valid logprobs are zeros.
+            - logprobs_valid is a 1-d boolean tesor with flags indicating whether the corresponding
+            element in the logprobs is a valid logprob (True) or not (False).
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:735]: When multiple variables are returned (indeed, as a tuple), the way to list them in the docstrings is one by one. I will make this change myself.

Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py]: https://github.com/alexhernandezgarcia/gflownet/pull/396/commits/4a67c0d08362bbdc166a753d959e9fcc4c95f211

File: gflownet/utils/batch.py (hunk 4 above)
Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py:749]: > No, it will not :) It will call inside the torch.stack over all the one-element tensors inside the logprobs_clean and this operation preserves the connection to the computational graphs
Uhm ok, I had done a quick independent test and I could not see the gradients any more. But it was because I was not calling stack like you noted.

Then my other suggestion would work, right? Namely:
```
logprobs_clean = [logp if logp not None else 0.0 for logp in logprobs]
```

File: gflownet/utils/batch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py]: Thanks!!!

File: gflownet/utils/batch.py [hunk 9]
```diff
@@ -942,6 +997,7 @@ def get_masks_forward(
         self,
         of_parents: bool = False,
         force_recompute: bool = False,
+        index: Optional[Union[List, Tuple, TensorType, npt.NDArray]] = None,
```
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py]: Thanks! :)

File: gflownet/utils/batch.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py]: Thank you for adding this :)

File: gflownet/utils/batch.py (hunk 4 above)
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py:749]: Unfortunately, this won't work either, because some of the logp elements are one-element tensors, so tfloat will try to stack them with the floats (zeros which you put here) and throw this error: 
Screenshot 2025-09-19 at 18 10 49

Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py:749]: Ok, my final attempt:
```
logprobs_clean = [logp if logp not None else torch.tensor(0.0) for logp in logprobs]
```
I guess using tfloat instead of just torch.tensor...

Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py:749]: That would work, but I also need to update `logprobs_valid` simultaneously as I'm creating `logprobs_clean`

Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py:749]: Ok let's merge it and simplify this later.

**Comment by AlexandraVolokhova** [comment]
@alexhernandezgarcia could you approve this PR? I cannot merge it myself w/o an approval

**Review by alexhernandezgarcia — APPROVED** [review]
OK to merge
